src/addon/file_server/mod.rs

```diff
@@ -64,8 +64,9 @@ impl<'a> FileServer {
 
         if let Some(path_and_query) = uri_parts.path_and_query {
             let path = path_and_query.path();
+            let path = path.replace("%20", " ");
```

Comment:
As of today, spaces are the only characters causing this issue, so I will start by just replacing them instead of applying a full change to the whole path. The `path` here is then used to resolve the filename in the scoped file system.

Reply:
Just replacing spaces with `%20` and vice versa doesn't solve the general problem. For example, things still break with a file called `?test.txt`.

I think the safest thing to do here is to pull in the [urlencoding](https://docs.rs/urlencoding/2.1.0/urlencoding/) crate or something similar to get a reasonably-edge-case-free transformation.

Reply:
Oh, I missed the comment about not wanting to encode slashes... that's kinda tricky.

You probably would have to url-encode each path segment separately, but I think that gets even messier if you want to support filenames that aren't valid UTF-8.

Reply:
Yeah, I started to think about those scenarios as well. Theres many static file servers out there, I think I will have to dig further into best practices for this use case on URLs and find a better solution.

Perhaps another solution could be: Splitting by slashes, encode each resulting element and then join with slashes.

What do you think?

Reply:
Yeah, that's what I would go for. I ended up just writing an implementation of this and turns it it's way uglier than I thought it would be.

My general strategy was:

 - Encode each segment that is valid unicode with standard UTF-8 percent-encoding
 - Encode any segments that are not valid unicode in the platform's native representation

That "in the platform's native representation" part is painful though, because windows paths are UTF-16. Rust's standard library abstracts this away pretty well, but as far as I can tell the only thing to do when they aren't valid unicode is to handle the UTF-16 nonsense explicitly.

I wrote up a rough implementation of this using the [percent_encoding](https://docs.rs/percent-encoding/) crate (I abandoned the urlencoding crate due to [this issue](https://github.com/kornelski/rust_urlencoding/issues/4)):

```rust
use percent_encoding::{utf8_percent_encode, percent_encode, AsciiSet, NON_ALPHANUMERIC};

const PERCENT_ENCODE_SET: &AsciiSet = &NON_ALPHANUMERIC
    .remove(b'-')
    .remove(b'_')
    .remove(b'.')
    .remove(b'~');

let file_path: PathBuf = ...;

let url_path = file_path.iter().flat_map(|component| {
    let segment = match component.to_str() {
        Some(component) => utf8_percent_encode(component, PERCENT_ENCODE_SET),
        None => {
            let bytes = {
                #[cfg(windows)]
                {
                    // Not having [T; N]: IntoIterator sucks
                    let mut bytes = Vec::with_capacity(component.len() * 2);
                    for wc in component.encode_wide() {
                        let wc = wc.to_be_bytes();
                        bytes.push(wc[0]);
                        bytes.push(wc[1]);
                    }
                    bytes
                }
                #[cfg(not(windows))]
                component.as_bytes()
            };
            percent_encode(&bytes, PERCENT_ENCODE_SET)
        }
    };
    std::iter::once("/").chain(segment)
}).collect::<String>();
```

Going the other direction has another fun windows-specific case to handle, where the percent-decoded path segment doesn't break into pairs:

```rust
let segments = url.path_segments().ok_or(Err(...))?;
let file_path = segments.map(|segment| {
    match percent_decode_str(segment).decode_utf8() {
        Ok(component) => Ok(OsStr::new(&component[..]).to_owned()),
        Err(_) => {
            let bytes = percent_decode_str(segment).collect::<Vec<u8>>();
            #[cfg(windows)]
            {
                let wide_bytes = bytes.iter().chunks(2).map(|slice| {
                    match slice.try_into::<[u8; 2]>() {
                        Ok(pair) => Ok(u16::from_be_bytes(pair)),
                        // This path segment decodes as an odd number of bytes,
                        // which means we can't convert it to an OsString on
                        // windows, and it will never represent a valid path.
                        // Should probably just return a 404 in this situation.
                        Err(err) => Err(...)
                    }
                }).collect::<Result<Vec<u16>, _>>()?;
                Ok(OsString::from_wide(&wide_bytes))
            }
            #[cfg(not(windows))]
            Ok(OsStr::from_bytes(&bytes).to_owned())
        }
    }
}).collect::<Result<PathBuf, _>>();
```

I know `OsString` uses WTF-8 internally, and my understanding is that if `OsString`/`OsStr` exposed that representation directly as a byte slice we could just use that without any platform-specific shenanigans. Unfortunately, afaict no such API exists. I'd be happy to find out I'm wrong on this one though.

Reply:
@Benjamin-L thanks so much for these snippets and the explanation, its very helpful to be in context!
I agree with you on the approach you propose here, if you want to give it a shot in a PR please do so, I'm glad to work with you on this.

I think I will have some time to take care of this between later today and tomorrow.

Thanks again!

Reply:
I'm pretty busy the next few days, and probably don't have the time to familiarize myself with the `http-server` codebase enough to put together a PR. I'd be happy to review one though.

Reply:
No worries! I will keep pushing on this PR soon and ping you!